[Bug]: Error when importing aws_glue_partition

### Terraform Core Version

1.9.3

### AWS Provider Version

1.

### Affected Resource(s)

aws_glue_partition

When importing a partition the import throws an unexpected error

```console
│ Error: setting storage_descriptor: Invalid address to set: []string{"storage_descriptor", "0", "additional_locations"}
```

### Expected Behavior

Resource should be imported without this error

### Actual Behavior

The resource is not imported and you get this error:

```console
│ Error: setting storage_descriptor: Invalid address to set: []string{"storage_descriptor", "0", "additional_locations"}
```

### Relevant Error/Panic Output Snippet

```console
│ Error: setting storage_descriptor: Invalid address to set: []string{"storage_descriptor", "0", "additional_locations"}
```


### Terraform Configuration Files

```hcl
import {
  to = aws_glue_partition.p-566972129213__sechub__brewed_sechub2_recipe_output__sechub2_29Mar2021_1617005427301
  id = "566972129213:sechub:brewed_sechub2_recipe_output:sechub2_29Mar2021_1617005427301"
}
```

### Steps to Reproduce

import.tf:

```hcl
import {
  to = aws_glue_partition.p-566972129213__sechub__brewed_sechub2_recipe_output__sechub2_29Mar2021_1617005427301
  id = "566972129213:sechub:brewed_sechub2_recipe_output:sechub2_29Mar2021_1617005427301"
}
```

```console
terraform plan -generate-config-out=resources.out

│ Error: setting storage_descriptor: Invalid address to set: []string{"storage_descriptor", "0", "additional_locations"}
```

& the resource is not imported.



Here is the partition configuration:

```json
awsandy@bcd0749e47ec my-aws-samples % aws glue get-partitions --database-name sechub --table-name brewed_sechub2_recipe_output
{
    "Partitions": [
        {
            "Values": [
                "sechub2_29Mar2021_1617005427301"
            ],
            "DatabaseName": "sechub",
            "TableName": "brewed_sechub2_recipe_output",
            "CreationTime": "2021-03-29T11:18:50+01:00",
            "LastAccessTime": "2021-03-29T11:18:50+01:00",
            "StorageDescriptor": {
                "Columns": [
                    {
                        "Name": "awsaccountid",
                        "Type": "string"
                    },
                    {
                        "Name": "compliance_status",
                        "Type": "string"
                    },
                    {
                        "Name": "compliance_statusreasons",
                        "Type": "array<struct<Description:string,ReasonCode:string>>"
                    },
                    {
                        "Name": "createdat",
                        "Type": "string"
                    },
                    {
                        "Name": "description",
                        "Type": "string"
                    },
                    {
                        "Name": "firstobservedat",
                        "Type": "string"
                    },
                    {
                        "Name": "generatorid",
                        "Type": "string"
                    },
                    {
                        "Name": "id",
                        "Type": "string"
                    },
                    {
                        "Name": "lastobservedat",
                        "Type": "string"
                    },
                    {
                        "Name": "productarn",
                        "Type": "string"
                    },
                    {
                        "Name": "productfields_recommendationurl",
                        "Type": "string"
                    },
                    {
                        "Name": "productfields_relatedawsresources:0/name",
                        "Type": "string"
                    },
                    {
                        "Name": "productfields_relatedawsresources:0/type",
                        "Type": "string"
                    },
                    {
                        "Name": "productfields_relatedawsresources:1/name",
                        "Type": "string"
                    },
                    {
                        "Name": "productfields_relatedawsresources:1/type",
                        "Type": "string"
                    },
                    {
                        "Name": "productfields_ruleid",
                        "Type": "string"
                    },
                    {
                        "Name": "productfields_standardscontrolarn",
                        "Type": "string"
                    },
                    {
                        "Name": "productfields_standardsguidearn",
                        "Type": "string"
                    },
                    {
                        "Name": "productfields_standardsguidesubscriptionarn",
                        "Type": "string"
                    },
                    {
                        "Name": "productfields_aws/securityhub/companyname",
                        "Type": "string"
                    },
                    {
                        "Name": "productfields_aws/securityhub/findingid",
                        "Type": "string"
                    },
                    {
                        "Name": "productfields_aws/securityhub/productname",
                        "Type": "string"
                    },
                    {
                        "Name": "productfields_aws/securityhub/annotation",
                        "Type": "string"
                    },
                    {
                        "Name": "recordstate",
                        "Type": "string"
                    },
                    {
                        "Name": "remediation_recommendation",
                        "Type": "struct<Text:string,Url:string>"
                    },
                    {
                        "Name": "awsresources",
                        "Type": "struct<Details:struct<AwsCloudTrailTrail:struct<CloudWatchLogsLogGroupArn:string,CloudWatchLogsRoleArn:string,HasCustomEventSelectors:boolean,HomeRegion:string,IncludeGlobalServiceEvents:boolean,IsMultiRegionTrail:boolean,IsOrganizationTrail:boolean,KmsKeyId:string,LogFileValidationEnabled:boolean,Name:string,S3BucketName:string,S3KeyPrefix:string>,AwsEc2SecurityGroup:struct<GroupId:string,GroupName:string,IpPermissions:array<struct<FromPort:bigint,IpProtocol:string,IpRanges:array<struct<CidrIp:string>>,Ipv6Ranges:array<struct<CidrIpv6:string>>,ToPort:bigint,UserIdGroupPairs:array<struct<GroupId:string,UserId:string>>>>,IpPermissionsEgress:array<struct<FromPort:bigint,IpProtocol:string,IpRanges:array<struct<CidrIp:string>>,Ipv6Ranges:array<struct<CidrIpv6:string>>,ToPort:bigint,UserIdGroupPairs:array<struct<GroupId:string,UserId:string>>>>,OwnerId:string,VpcId:string>,AwsEc2Vpc:struct<CidrBlockAssociationSet:array<struct<AssociationId:string,CidrBlock:string,CidrBlockState:string>>,DhcpOptionsId:string,State:string>,AwsIamPolicy:struct<AttachmentCount:bigint,CreateDate:string,DefaultVersionId:string,Description:string,IsAttachable:boolean,Path:string,PermissionsBoundaryUsageCount:bigint,PolicyId:string,PolicyName:string,PolicyVersionList:array<struct<CreateDate:string,IsDefaultVersion:boolean,VersionId:string>>,UpdateDate:string>,AwsIamUser:struct<AttachedManagedPolicies:array<struct<PolicyArn:string,PolicyName:string>>,CreateDate:string,GroupList:array<string>,Path:string,UserId:string,UserName:string,UserPolicyList:array<struct<PolicyName:string>>>,AwsKmsKey:struct<AWSAccountId:string,CreationDate:bigint,Description:string,KeyId:string,KeyManager:string,KeyState:string,Origin:string>>,Id:string,Partition:string,Region:string,Type:string>"
                    },
                    {
                        "Name": "schemaversion",
                        "Type": "string"
                    },
                    {
                        "Name": "severity_label",
                        "Type": "string"
                    },
                    {
                        "Name": "severity_normalized",
                        "Type": "bigint"
                    },
                    {
                        "Name": "severity_original",
                        "Type": "string"
                    },
                    {
                        "Name": "severity_product",
                        "Type": "bigint"
                    },
                    {
                        "Name": "title",
                        "Type": "string"
                    },
                    {
                        "Name": "types",
                        "Type": "array<string>"
                    },
                    {
                        "Name": "updatedat",
                        "Type": "string"
                    },
                    {
                        "Name": "workflow_status",
                        "Type": "string"
                    },
                    {
                        "Name": "workflowstate",
                        "Type": "string"
                    }
                ],
                "Location": "s3://at-hmrc1/sechub2-recipe-output/sechub2_29Mar2021_1617005427301/",
                "InputFormat": "org.apache.hadoop.hive.ql.io.parquet.MapredParquetInputFormat",
                "OutputFormat": "org.apache.hadoop.hive.ql.io.parquet.MapredParquetOutputFormat",
                "Compressed": false,
                "NumberOfBuckets": -1,
                "SerdeInfo": {
                    "SerializationLibrary": "org.apache.hadoop.hive.ql.io.parquet.serde.ParquetHiveSerDe",
                    "Parameters": {
                        "serialization.format": "1"
                    }
                },
                "BucketColumns": [],
                "SortColumns": [],
                "Parameters": {
                    "sizeKey": "570207",
                    "objectCount": "1",
                    "recordCount": "869",
                    "averageRecordSize": "617",
                    "compressionType": "none",
                    "classification": "parquet",
                    "typeOfData": "file"
                },
                "StoredAsSubDirectories": false
            },
            "Parameters": {},
            "CatalogId": "566972129213"
        }
    ]
}
```


### Debug Output

_No response_

### Panic Output

_No response_

### Important Factoids

_No response_

### References

_No response_

### Would you like to implement a fix?

None

Using `TF_LOG=debug` shows the following, note `tf_rpc=ReadResource`:

```
2025-02-03T10:21:51.441+0200 [ERROR] provider.terraform-provider-aws_v5.84.0_x5: [ERROR] setting state: Invalid address to set: []string{"storage_descriptor", "0", "additional_locations"}
2025-02-03T10:21:51.443+0200 [ERROR] provider.terraform-provider-aws_v5.84.0_x5: Response contains error diagnostic: diagnostic_summary="setting storage_descriptor: Invalid address to set: []string{\"storage_descriptor\", \"0\", \"additional_locations\"}" @caller=github.com/hashicorp/terraform-plugin-go@v0.25.0/tfprotov5/internal/diag/diagnostics.go:58 @module=sdk.proto diagnostic_severity=ERROR tf_req_id=ea629bc8-82a6-82aa-b800-56693711b92c tf_resource_type=aws_glue_partition diagnostic_detail="" tf_proto_version=5.7 tf_provider_addr=registry.terraform.io/hashicorp/aws tf_rpc=ReadResource timestamp="2025-02-03T10:21:51.443+0200"
```

It looks like https://github.com/nikhil-goenka/terraform-provider-aws/commit/bc86ca4ed0e3e529161e594646492895abf20b14#diff-5e4d893235d1f275a3d26ed5f4c5c22f4d2ef57dfbeb30bb29a4c6b15cca9121R915 modified the `flattenStorageDescriptor()` function in `catalog_table.go` to return a `additional_locations` key in the `[]map[string]interface{}` return value.

https://github.com/nikhil-goenka/terraform-provider-aws/blob/c428d2cb510d33fcb603d7e6e22d1370416f642a/internal/service/glue/partition.go#L267-L269 however, the `aws_glue_partition` resource in `partition.go` uses the same `flattenStorageDescriptor()` function from `catalog_table.go` in its resource read function to set the `storage_descriptor` data, which fails because of the `additional_locations` key that is not defined in the `aws_glue_partition` resource schema.

I think this is thus a regression in https://github.com/hashicorp/terraform-provider-aws/pull/37891 -> terraform-provider-aws 5.55 that breaks all use of `aws_glue_partition`. Ping @nikhil-goenka ?

--- a/.changelog/41434.txt
+++ b/.changelog/41434.txt
@@ -1,3 +1,3 @@
 ```release-note:enhancement
-resource/aws_glue_partition: Add `additional_locations` argument in `storage_descriptor`
+resource/aws_glue_partition: Add `storage_descriptor.additional_locations` argument
 ```

--- a/.changelog/41434.txt
+++ b/.changelog/41434.txt
@@ -1,0 +1,3 @@
+```release-note:enhancement
+resource/aws_glue_partition: Add `additional_locations` argument in `storage_descriptor`
+```

--- a/internal/service/glue/partition.go
+++ b/internal/service/glue/partition.go
@@ -66,6 +66,11 @@ func ResourcePartition() *schema.Resource {
 				MaxItems: 1,
 				Elem: &schema.Resource{
 					Schema: map[string]*schema.Schema{
+						"additional_locations": {
+							Type:     schema.TypeList,
+							Optional: true,
+							Elem:     &schema.Schema{Type: schema.TypeString},
+						},
 						"bucket_columns": {
 							Type:     schema.TypeList,
 							Optional: true,

--- a/internal/service/glue/partition_test.go
+++ b/internal/service/glue/partition_test.go
@@ -189,6 +189,63 @@ func TestAccGluePartition_storageDescriptorBasic(t *testing.T) {
 	})
 }
 
+func TestAccGluePartition_storageDescriptorAdditionalLocations(t *testing.T) {
+	ctx := acctest.Context(t)
+	rName := sdkacctest.RandomWithPrefix(acctest.ResourcePrefix)
+	parValue := sdkacctest.RandString(10)
+	parValue2 := sdkacctest.RandString(10)
+	parValue3 := sdkacctest.RandString(10)
+	resourceName := "aws_glue_partition.test"
+
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck:                 func() { acctest.PreCheck(ctx, t) },
+		ErrorCheck:               acctest.ErrorCheck(t, names.GlueServiceID),
+		ProtoV5ProviderFactories: acctest.ProtoV5ProviderFactories,
+		CheckDestroy:             testAccCheckPartitionDestroy(ctx),
+		Steps: []resource.TestStep{
+			{
+				Config: testAccPartitionConfig_storageDescriptorAdditionalLocationsSingle(rName, parValue),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckPartitionExists(ctx, resourceName),
+					acctest.CheckResourceAttrAccountID(ctx, resourceName, names.AttrCatalogID),
+					resource.TestCheckResourceAttr(resourceName, names.AttrDatabaseName, rName),
+					resource.TestCheckResourceAttr(resourceName, "storage_descriptor.0.additional_locations.#", "1"),
+					resource.TestCheckResourceAttr(resourceName, "storage_descriptor.0.additional_locations.0", parValue),
+					resource.TestCheckResourceAttrSet(resourceName, names.AttrCreationTime),
+				),
+			},
+			{
+				ResourceName:      resourceName,
+				ImportState:       true,
+				ImportStateVerify: true,
+			},
+			{
+				Config: testAccPartitionConfig_storageDescriptorAdditionalLocationsDouble(rName, parValue, parValue2),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckPartitionExists(ctx, resourceName),
+					acctest.CheckResourceAttrAccountID(ctx, resourceName, names.AttrCatalogID),
+					resource.TestCheckResourceAttr(resourceName, names.AttrDatabaseName, rName),
+					resource.TestCheckResourceAttr(resourceName, "storage_descriptor.0.additional_locations.#", "2"),
+					resource.TestCheckResourceAttr(resourceName, "storage_descriptor.0.additional_locations.0", parValue),
+					resource.TestCheckResourceAttr(resourceName, "storage_descriptor.0.additional_locations.1", parValue2),
+					resource.TestCheckResourceAttrSet(resourceName, names.AttrCreationTime),
+				),
+			},
+			{
+				Config: testAccPartitionConfig_storageDescriptorAdditionalLocationsSingle(rName, parValue3),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckPartitionExists(ctx, resourceName),
+					acctest.CheckResourceAttrAccountID(ctx, resourceName, names.AttrCatalogID),
+					resource.TestCheckResourceAttr(resourceName, names.AttrDatabaseName, rName),
+					resource.TestCheckResourceAttr(resourceName, "storage_descriptor.0.additional_locations.#", "1"),
+					resource.TestCheckResourceAttr(resourceName, "storage_descriptor.0.additional_locations.0", parValue3),
+					resource.TestCheckResourceAttrSet(resourceName, names.AttrCreationTime),
+				),
+			},
+		},
+	})
+}
+
 func TestAccGluePartition_disappears(t *testing.T) {
 	ctx := acctest.Context(t)
 	rName := sdkacctest.RandomWithPrefix(acctest.ResourcePrefix)
@@ -513,4 +570,64 @@ resource "aws_glue_partition" "test" {
   }
 }
 `, parValue)
+}
+
+func testAccPartitionConfig_storageDescriptorAdditionalLocationsSingle(rName, additionalLocation1 string) string {
+	return testAccPartitionConfigBase(rName) +
+		fmt.Sprintf(`
+resource "aws_glue_partition" "test" {
+  database_name    = aws_glue_catalog_database.test.name
+  table_name       = aws_glue_catalog_table.test.name
+  partition_values = ["2025"]
+
+  storage_descriptor {
+    additional_locations = [%[1]q]
+    location             = "my_location"
+    input_format         = "org.apache.hadoop.mapred.TextInputFormat"
+    output_format        = "org.apache.hadoop.hive.ql.io.HiveIgnoreKeyTextOutputFormat"
+    compressed           = false
+    number_of_buckets    = 1
+
+    ser_de_info {
+      name = "ser_de_name"
+
+      parameters = {
+        "example" = "test"
+      }
+
+      serialization_library = "org.apache.hadoop.hive.serde2.columnar.ColumnarSerDe"
+    }
+  }
+}
+`, additionalLocation1)
+}
+
+func testAccPartitionConfig_storageDescriptorAdditionalLocationsDouble(rName, additionalLocation1, additionalLocation2 string) string {
+	return testAccPartitionConfigBase(rName) +
+		fmt.Sprintf(`
+resource "aws_glue_partition" "test" {
+  database_name    = aws_glue_catalog_database.test.name
+  table_name       = aws_glue_catalog_table.test.name
+  partition_values = ["2025"]
+
+  storage_descriptor {
+    additional_locations = [%[1]q, %[2]q]
+    location             = "my_location"
+    input_format         = "org.apache.hadoop.mapred.TextInputFormat"
+    output_format        = "org.apache.hadoop.hive.ql.io.HiveIgnoreKeyTextOutputFormat"
+    compressed           = false
+    number_of_buckets    = 1
+
+    ser_de_info {
+      name = "ser_de_name"
+
+      parameters = {
+        "example" = "test"
+      }
+
+      serialization_library = "org.apache.hadoop.hive.serde2.columnar.ColumnarSerDe"
+    }
+  }
+}
+`, additionalLocation1, additionalLocation2)
 }

--- a/internal/service/glue/partition_test.go
+++ b/internal/service/glue/partition_test.go
@@ -129,6 +129,66 @@ func TestAccGluePartition_parameters(t *testing.T) {
 	})
 }
 
+func TestAccGluePartition_storageDescriptorBasic(t *testing.T) {
+	ctx := acctest.Context(t)
+	rName := sdkacctest.RandomWithPrefix(acctest.ResourcePrefix)
+	parValue := sdkacctest.RandString(10)
+	parValue2 := sdkacctest.RandString(10)
+	resourceName := "aws_glue_partition.test"
+
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck:                 func() { acctest.PreCheck(ctx, t) },
+		ErrorCheck:               acctest.ErrorCheck(t, names.GlueServiceID),
+		ProtoV5ProviderFactories: acctest.ProtoV5ProviderFactories,
+		CheckDestroy:             testAccCheckPartitionDestroy(ctx),
+		Steps: []resource.TestStep{
+			{
+				Config: testAccPartitionConfig_storageDescriptorBasic(rName, parValue),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckPartitionExists(ctx, resourceName),
+					acctest.CheckResourceAttrAccountID(ctx, resourceName, names.AttrCatalogID),
+					resource.TestCheckResourceAttr(resourceName, names.AttrDatabaseName, rName),
+					resource.TestCheckResourceAttr(resourceName, "storage_descriptor.0.location", "my_location"),
+					resource.TestCheckResourceAttr(resourceName, "storage_descriptor.0.input_format", "org.apache.hadoop.mapred.TextInputFormat"),
+					resource.TestCheckResourceAttr(resourceName, "storage_descriptor.0.output_format", "org.apache.hadoop.hive.ql.io.HiveIgnoreKeyTextOutputFormat"),
+					resource.TestCheckResourceAttr(resourceName, "storage_descriptor.0.compressed", acctest.CtFalse),
+					resource.TestCheckResourceAttr(resourceName, "storage_descriptor.0.number_of_buckets", "1"),
+					resource.TestCheckResourceAttr(resourceName, "storage_descriptor.0.ser_de_info.#", "1"),
+					resource.TestCheckResourceAttr(resourceName, "storage_descriptor.0.ser_de_info.0.name", "ser_de_name"),
+					resource.TestCheckResourceAttr(resourceName, "storage_descriptor.0.ser_de_info.0.parameters.%", "1"),
+					resource.TestCheckResourceAttr(resourceName, "storage_descriptor.0.ser_de_info.0.parameters.example", parValue),
+					resource.TestCheckResourceAttr(resourceName, "storage_descriptor.0.ser_de_info.0.serialization_library", "org.apache.hadoop.hive.serde2.columnar.ColumnarSerDe"),
+					resource.TestCheckResourceAttrSet(resourceName, names.AttrCreationTime),
+				),
+			},
+			{
+				ResourceName:      resourceName,
+				ImportState:       true,
+				ImportStateVerify: true,
+			},
+			{
+				Config: testAccPartitionConfig_storageDescriptorBasic(rName, parValue2),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckPartitionExists(ctx, resourceName),
+					acctest.CheckResourceAttrAccountID(ctx, resourceName, names.AttrCatalogID),
+					resource.TestCheckResourceAttr(resourceName, names.AttrDatabaseName, rName),
+					resource.TestCheckResourceAttr(resourceName, "storage_descriptor.0.location", "my_location"),
+					resource.TestCheckResourceAttr(resourceName, "storage_descriptor.0.input_format", "org.apache.hadoop.mapred.TextInputFormat"),
+					resource.TestCheckResourceAttr(resourceName, "storage_descriptor.0.output_format", "org.apache.hadoop.hive.ql.io.HiveIgnoreKeyTextOutputFormat"),
+					resource.TestCheckResourceAttr(resourceName, "storage_descriptor.0.compressed", acctest.CtFalse),
+					resource.TestCheckResourceAttr(resourceName, "storage_descriptor.0.number_of_buckets", "1"),
+					resource.TestCheckResourceAttr(resourceName, "storage_descriptor.0.ser_de_info.#", "1"),
+					resource.TestCheckResourceAttr(resourceName, "storage_descriptor.0.ser_de_info.0.name", "ser_de_name"),
+					resource.TestCheckResourceAttr(resourceName, "storage_descriptor.0.ser_de_info.0.parameters.%", "1"),
+					resource.TestCheckResourceAttr(resourceName, "storage_descriptor.0.ser_de_info.0.parameters.example", parValue2),
+					resource.TestCheckResourceAttr(resourceName, "storage_descriptor.0.ser_de_info.0.serialization_library", "org.apache.hadoop.hive.serde2.columnar.ColumnarSerDe"),
+					resource.TestCheckResourceAttrSet(resourceName, names.AttrCreationTime),
+				),
+			},
+		},
+	})
+}
+
 func TestAccGluePartition_disappears(t *testing.T) {
 	ctx := acctest.Context(t)
 	rName := sdkacctest.RandomWithPrefix(acctest.ResourcePrefix)
@@ -424,4 +484,33 @@ resource "aws_glue_partition" "test" {
   partition_values = ["%[2]s", "%[3]s"]
 }
 `, rName, parValue, parValue2)
+}
+
+func testAccPartitionConfig_storageDescriptorBasic(rName, parValue string) string {
+	return testAccPartitionConfigBase(rName) +
+		fmt.Sprintf(`
+resource "aws_glue_partition" "test" {
+  database_name    = aws_glue_catalog_database.test.name
+  table_name       = aws_glue_catalog_table.test.name
+  partition_values = ["2025"]
+
+  storage_descriptor {
+    location          = "my_location"
+    input_format      = "org.apache.hadoop.mapred.TextInputFormat"
+    output_format     = "org.apache.hadoop.hive.ql.io.HiveIgnoreKeyTextOutputFormat"
+    compressed        = false
+    number_of_buckets = 1
+
+    ser_de_info {
+      name = "ser_de_name"
+
+      parameters = {
+        "example" = %[1]q
+      }
+
+      serialization_library = "org.apache.hadoop.hive.serde2.columnar.ColumnarSerDe"
+    }
+  }
+}
+`, parValue)
 }

--- a/website/docs/r/glue_partition.html.markdown
+++ b/website/docs/r/glue_partition.html.markdown
@@ -32,6 +32,7 @@ This resource supports the following arguments:
 
 ##### storage_descriptor
 
+* `additional_locations` - (Optional) List of locations that point to the path where a Delta table is located.
 * `columns` - (Optional) A list of the [Columns](#column) in the table.
 * `location` - (Optional) The physical location of the table. By default this takes the form of the warehouse location, followed by the database location in the warehouse, followed by the table name.
 * `input_format` - (Optional) The input format: SequenceFileInputFormat (binary), or TextInputFormat, or a custom format.

